src/System.Composition.Convention/src/Microsoft/Internal/Lock.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Internal
+{
+    internal sealed class Lock : IDisposable
+    {
+#if FEATURE_SLIMLOCK
+        private ReaderWriterLockSlim _thisLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+        private int _isDisposed = 0;
+        public void EnterReadLock()
+        {
+            this._thisLock.EnterReadLock();
+        }
+
+        public void EnterWriteLock()
+        {
+            this._thisLock.EnterWriteLock();
+        }
+
+        public void ExitReadLock()
+        {
+            this._thisLock.ExitReadLock();
+        }
+
+        public void ExitWriteLock()
+        {
+            this._thisLock.ExitWriteLock();
+        }
+
+        public void Dispose()
+        {
+            if (Interlocked.CompareExchange(ref this._isDisposed, 1, 0) == 0)
```

Comment:
Why protect against multiple Dispose calls?  Is the concern that code might be using Dispose concurrently?  If so, that seems like poorly written consuming code.  Otherwise, shouldn't Dispose allow for multiple calls to Dispose, where subsequent calls are nops?